File: packages/vela-sdk-ts/src/engine/prompt-builder.ts
/**
 * Prompt assembly and template resolution for workflow steps.
 *
 * Pure logic -- no store dependency. Handles template resolution,
 * progress indicators, resource assembly, and CTAs.
 */

import type { WorkflowRunState } from "./types.js";
import type { ResourceDefinition } from "../schemas/resource.js";
import type {
  AnyStepDefinition,
  WorkflowDefinition,
} from "../schemas/workflow.js";
import { getLocale, type Locale } from "../locale/locale.js";

/** Resolves a resource ref ID to its definition. */
export type ResourceResolver = (refId: string) => ResourceDefinition | undefined;

/**
 * Resolves a run's `projectId` to arbitrary project data, exposed in
 * templates as `{{project.x}}`. The SDK ships no implementation — the
 * embedding app supplies one (e.g. backed by its own project store).
 */
export type ProjectDataResolver = (projectId: string) => Record<string, unknown> | undefined;

// ---------------------------------------------------------------------------
// PromptBuilder
// ---------------------------------------------------------------------------

export class PromptBuilder {
  /**
   * Build nested context dict for template resolution.
   *
   * Supports: {{params.X}}, {{steps.step_id.capture_key}}, {{state.key}}
   */
  static buildTemplateContext(
    workflowDef: WorkflowDefinition,
    run: WorkflowRunState,
    projectDataResolver?: ProjectDataResolver,
  ): Record<string, unknown> {
    const state = run.stateData;
    const params = run.params;

    // Build steps context: map step_id -> {capture_key: value}
    const stepsContext: Record<string, Record<string, unknown>> = {};
    for (const stepDef of workflowDef.steps) {
      const stepData: Record<string, unknown> = {};
      for (const cap of stepDef.capture) {
        if (cap.key in state) {
          stepData[cap.key] = state[cap.key];
        }
      }
      if (Object.keys(stepData).length > 0) {
        stepsContext[stepDef.id] = stepData;
      }
    }

    // {{resolved.x}} — the subset of params flagged `resolve: true` in the
    // workflow definition. Same values as `params.x`, just also reachable
    // under this name.
    const resolved: Record<string, unknown> = {};
    for (const pDef of workflowDef.params) {
      if (pDef.resolve && pDef.name in params) {
        resolved[pDef.name] = params[pDef.name];
      }
    }

    // {{project.x}} — populated only if the caller supplied a resolver.
    const project = run.projectId ? (projectDataResolver?.(run.projectId) ?? {}) : {};

    // {{fetch.x}} — results of the current step's `fetch` definitions,
    // recomputed each time a step is entered (see `resolveFetchesForStep`).
    const fetch = (state["_fetch"] as Record<string, unknown> | undefined) ?? {};

    return {
      params,
      steps: stepsContext,
      state,
      resolved,
      project,
      fetch,
    };
  }

  /** Resolve {{variable}} templates in text. */
  static resolveTemplates(text: string, context: Record<string, unknown>): string {
    return text.replace(/\{\{(.+?)\}\}/g, (_match, rawKey: string) => {
      const key = rawKey.trim();
      const keyParts = key.split(".");
      let value: unknown = context;
      for (const part of keyParts) {
        if (value !== null && typeof value === "object" && !Array.isArray(value)) {
          value = (value as Record<string, unknown>)[part];
          if (value === undefined) {
            return `{{${key}}}`;
          }
        } else {
          return `{{${key}}}`;
        }
      }
      return String(value);
    });
  }

  /**
   * Assemble resource sections for the prompt.
   *
   * Merges workflow-level and step-level resources (step wins on same ref).
   * Resources < 500 chars are inlined; others are listed as URI references.
   */
  static assembleResources(
    workflowDef: WorkflowDefinition,
    step: AnyStepDefinition,
    resourceResolver: ResourceResolver,
    forceInline = false,
    locale?: Locale,
  ): string[] {
    const loc = locale ?? getLocale();

    // Merge: workflow-level first, step-level overrides
    const merged = new Map<string, { ref: string; inline?: boolean | null }>();
    for (const ref of workflowDef.resources) {
      merged.set(ref.ref, ref);
    }
    for (const ref of step.resources) {
      merged.set(ref.ref, ref);
    }

    if (merged.size === 0) {
      return [];
    }

    const inlineParts: string[] = [];
    const referenceParts: string[] = [];

    for (const [, resRef] of merged) {
      const resource = resourceResolver(resRef.ref);
      if (!resource) {
        continue;
      }

      // Determine inline vs reference
      // Delegate steps always inline — the subagent is a separate session
      // and cannot load resources on demand.
      let shouldInline = forceInline || resRef.inline;
      if (shouldInline == null) {
        shouldInline = resource.content.length < 500;
      }

      if (shouldInline) {
        inlineParts.push(`### ${resource.name}`);
        inlineParts.push(resource.content);
        inlineParts.push("");
      } else {
        const uri =
          resource.uri_pattern ?? `vela://${resource.type}/${resource.id}`;
        const desc = resource.description ? ` — ${resource.description}` : "";
        referenceParts.push(`- \`${uri}\`${desc}`);
      }
    }

    const parts: string[] = [];
    if (inlineParts.length > 0) {
      parts.push(...inlineParts);
    }
    if (referenceParts.length > 0) {
      parts.push(loc.engineResourcesHeading);
      parts.push(...referenceParts);
      parts.push(loc.engineResourcesLoadHint);
    }

    return parts;
  }

  /**
   * Assemble the prompt for a step.
   *
   * Includes progress overview, depends_on context, resources, step prompt,
   * capture info, and CTA.
   */
  assemblePrompt(
    workflowDef: WorkflowDefinition,
    run: WorkflowRunState,
    step: AnyStepDefinition,
    resourceResolver?: ResourceResolver,
    locale?: Locale,
    projectDataResolver?: ProjectDataResolver,
  ): string {
    const loc = locale ?? getLocale();
    const state = run.stateData;
    const context = PromptBuilder.buildTemplateContext(workflowDef, run, projectDataResolver);

    const parts: string[] = [];

    // Header with step name
    const stepName = step.name ?? step.id;
    parts.push(`## ${workflowDef.name} — ${stepName}`);
    parts.push("");

    // Progress overview
    parts.push(loc.engineProgressHeading);
    for (const s of workflowDef.steps) {
      const sName = s.name ?? s.id;
      if (s.id === step.id) {
        parts.push(loc.engineProgressCurrentLine.replace("{step_name}", sName));
      } else if (s.capture.some((cap) => cap.key in state)) {
        parts.push(`- ~~${sName}~~ ✓`);
      } else {
        parts.push(`- ${sName}`);
      }
    }
    parts.push("");

    // depends_on context
    if (step.depends_on.length > 0) {
      parts.push(loc.engineDependsOnHeading);
      for (const dep of step.depends_on) {
        for (const field of dep.fields) {
          const value = state[field] ?? loc.engineNotCaptured;
          parts.push(`- **${field}**: ${value}`);
        }
      }
      parts.push("");
    }

    // Resources — force inline for delegate steps (subagent has no resource access)
    const isDelegate =
      step.type === "execute" && "delegate" in step && !!step.delegate;
    if (resourceResolver) {
      const resourceParts = PromptBuilder.assembleResources(
        workflowDef,
        step,
        resourceResolver,
        isDelegate,
        loc,
      );
      if (resourceParts.length > 0) {
        parts.push(...resourceParts);
        parts.push("");
      }
    }

    // Workflow-level tool requirements
    if (workflowDef.tools.length > 0) {
      parts.push(loc.engineToolsRequiredHeading);
      for (const t of workflowDef.tools) {
        const serverHint = t.server ? ` (${t.server})` : "";
        const descHint = t.description ? ` — ${t.description}` : "";
        const reqHint = t.required ? loc.engineRequiredTag : loc.engineOptionalTag;
        parts.push(`- **${t.name}**${serverHint}${descHint} ${reqHint}`);
      }
      parts.push("");
    }

    // Step-level tool hints
    if (step.tools.length > 0) {
      const toolList = step.tools.map((t) => `\`${t}\``).join(", ");
      parts.push(loc.engineStepToolsHeading);
      parts.push(loc.engineUseTheseTools.replace("{tools}", toolList));
      parts.push("");
    }

    // Step prompt with template resolution
    const prompt = PromptBuilder.resolveTemplates(step.prompt, context);
    parts.push(prompt);

    // Choice options
    if (step.type === "choice" && step.options.length > 0) {
      parts.push("");
      parts.push(loc.engineOptionsHeading);
      for (let i = 0; i < step.options.length; i++) {
        const opt = step.options[i];
        const desc = opt.description ? ` — ${opt.description}` : "";
        parts.push(`${i + 1}. **${opt.label}**${desc}`);
      }
    }

    // Capture info
    if (step.capture.length > 0) {
      parts.push("");
      const keys = step.capture.map((c) => c.key);
      parts.push(loc.engineCapturesHint.replace("{keys}", keys.join(", ")));
    }

    // CTA
    parts.push("");
    switch (step.type) {
      case "confirm":
        parts.push(loc.engineCtaConfirm);
        break;
      case "choice":
        parts.push(loc.engineCtaChoice);
        break;
      case "freeform":
        parts.push(loc.engineCtaFreeform);
        break;
      case "execute":
        if ("delegate" in step && step.delegate) {
          parts.push(loc.engineCtaDelegate.replace("{delegate}", step.delegate));
        } else {
          parts.push(loc.engineCtaExecute);
        }
        break;
      case "dialog":
        if (state["_dialog_phase"]) {
          parts.push(loc.engineCtaDialogContinue);
        } else {
          parts.push(loc.engineCtaDialogStart);
        }
        break;
      default:
        break;
    }

    return parts.join("\n");
  }
}

File: packages/vela-sdk/src/vela_sdk/engine/prompt_builder.py
"""Prompt assembly and template resolution for workflow steps."""

import re
from typing import Any, Callable, Optional

from vela_sdk.engine.types import WorkflowRunState
from vela_sdk.schemas.resource import ResourceDefinition
from vela_sdk.schemas.workflow import (
    AnyStepDefinition,
    StepType,
    WorkflowDefinition,
)


class PromptBuilder:
    """Assembles prompts for workflow steps.

    Pure logic — no store dependency. Handles template resolution,
    progress indicators, resource assembly, and CTAs.
    """

    @staticmethod
    def build_template_context(
        workflow_def: WorkflowDefinition,
        run: WorkflowRunState,
    ) -> dict[str, Any]:
        """Build nested context dict for template resolution.

        Supports: {{params.X}}, {{steps.step_id.capture_key}}, {{state.key}}
        """
        state = run.state_data
        params = run.params

        # Build steps context: map step_id -> {capture_key: value}
        steps_context: dict[str, dict[str, Any]] = {}
        for step_def in workflow_def.steps:
            step_data = {}
            for cap in step_def.capture:
                if cap.key in state:
                    step_data[cap.key] = state[cap.key]
            if step_data:
                steps_context[step_def.id] = step_data

        return {
            "params": params,
            "steps": steps_context,
            "state": state,
        }

    @staticmethod
    def resolve_templates(text: str, context: dict) -> str:
        """Resolve {{variable}} templates in text."""
        def replacer(match: re.Match[str]) -> str:
            key = match.group(1).strip()
            # Support dotted access: state.key
            key_parts = key.split(".")
            value: Any = context
            for part in key_parts:
                if isinstance(value, dict):
                    value = value.get(part, f"{{{{{key}}}}}")
                else:
                    return f"{{{{{key}}}}}"
            return str(value)

        return re.sub(r"\{\{(.+?)\}\}", replacer, text)

    @staticmethod
    def assemble_resources(
        workflow_def: WorkflowDefinition,
        step: AnyStepDefinition,
        resource_resolver: Callable[[str], Optional[ResourceDefinition]],
    ) -> list[str]:
        """Assemble resource sections for the prompt.

        Merges workflow-level and step-level resources (step wins on same ref).
        Resources < 500 chars are inlined; others are listed as URI references.
        """
        from vela_sdk.schemas.resource import ResourceReference
        # Merge: workflow-level first, step-level overrides
        merged: dict[str, ResourceReference] = {}
        for ref in workflow_def.resources:
            merged[ref.ref] = ref
        for ref in step.resources:
            merged[ref.ref] = ref

        if not merged:
            return []

        inline_parts: list[str] = []
        reference_parts: list[str] = []

        for ref_key, res_ref in merged.items():
            resource = resource_resolver(res_ref.ref)
            if not resource:
                continue

            # Determine inline vs reference
            should_inline = res_ref.inline
            if should_inline is None:
                should_inline = len(resource.content) < 500

            if should_inline:
                inline_parts.append(f"### {resource.name}")
                inline_parts.append(resource.content)
                inline_parts.append("")
            else:
                uri = resource.uri_pattern or f"vela://{resource.type.value}/{resource.id}"
                desc = f" — {resource.description}" if resource.description else ""
                reference_parts.append(f"- `{uri}`{desc}")

        parts: list[str] = []
        if inline_parts:
            parts.extend(inline_parts)
        if reference_parts:
            parts.append("### Available Resources")
            parts.extend(reference_parts)
            parts.append("*Lade mit `read_resource(\"URI\")` oder `vela_get_resource(id=\"...\")`.* ")

        return parts

    def assemble_prompt(
        self,
        workflow_def: WorkflowDefinition,
        run: WorkflowRunState,
        step: AnyStepDefinition,
        resource_resolver: Optional[Callable[[str], Optional[ResourceDefinition]]] = None,
    ) -> str:
        """Assemble the prompt for a step.

        Includes progress overview, depends_on context, resources, step prompt,
        capture info, and CTA.
        """
        state = run.state_data
        context = self.build_template_context(workflow_def, run)

        parts: list[str] = []

        # Header with step name
        step_name = step.name or step.id
        parts.append(f"## {workflow_def.name} — {step_name}")
        parts.append("")

        # Progress overview
        parts.append("### Fortschritt")
        for s in workflow_def.steps:
            s_name = s.name or s.id
            if s.id == step.id:
                parts.append(f"- **→ {s_name}** ← aktuell")
            elif any(cap.key in state for cap in s.capture):
                parts.append(f"- ~~{s_name}~~ ✓")
            else:
                parts.append(f"- {s_name}")
        parts.append("")

        # depends_on context: extract all field keys from DependsOnDefinition list
        if step.depends_on:
            parts.append("### Kontext aus vorherigen Steps:")
            for dep in step.depends_on:
                for field in dep.fields:
                    value = state.get(field, "(nicht erfasst)")
                    parts.append(f"- **{field}**: {value}")
            parts.append("")

        # Resources: merge workflow-level + step-level (step wins on same ref)
        if resource_resolver:
            resource_parts = self.assemble_resources(
                workflow_def, step, resource_resolver
            )
            if resource_parts:
                parts.extend(resource_parts)
                parts.append("")

        # Workflow-level tool requirements
        if workflow_def.tools:
            parts.append("### Benötigte externe Tools")
            for t in workflow_def.tools:
                server_hint = f" ({t.server})" if t.server else ""
                desc_hint = f" — {t.description}" if t.description else ""
                req_hint = "[erforderlich]" if t.required else "[optional]"
                parts.append(f"- **{t.name}**{server_hint}{desc_hint} {req_hint}")
            parts.append("")

        # Step-level tool hints
        if step.tools:
            tool_list = ", ".join(f"`{t}`" for t in step.tools)
            parts.append(f"### Tools für diesen Step")
            parts.append(f"Nutze folgende Tools: {tool_list}")
            parts.append("")

        # Step prompt with template resolution
        prompt = self.resolve_templates(step.prompt, context)
        parts.append(prompt)

        # Choice options
        if step.type == StepType.CHOICE and step.options:
            parts.append("")
            parts.append("### Optionen:")
            for i, opt in enumerate(step.options, 1):
                desc = f" — {opt.description}" if opt.description else ""
                parts.append(f"{i}. **{opt.label}**{desc}")

        # Capture info
        if step.capture:
            parts.append("")
            keys = [c.key for c in step.capture]
            parts.append(f"*Dieser Step erfasst: {', '.join(keys)}*")

        # CTA
        parts.append("")
        if step.type == StepType.CONFIRM:
            parts.append("**Bitte bestaetigen oder ablehnen.**")
        elif step.type == StepType.CHOICE:
            parts.append("**Bitte eine Option wählen.**")
        elif step.type == StepType.FREEFORM:
            parts.append("**Bitte Eingabe machen.**")
        elif step.type == StepType.EXECUTE:
            parts.append("**Ausführen, dann Abschluss bestaetigen.**")
        elif step.type == StepType.DIALOG:
            if state.get("_dialog_phase"):
                parts.append("**Dialog fortsetzen — aktuelle Phase bearbeiten.**")
            else:
                parts.append("**Dialog starten — advance aufrufen.**")

        return "\n".join(parts)
